### PR TITLE
Add cloud function for lat-lon to h3 conversion.

### PR DESCRIPTION
**Pull request by sylmak**
Add cloud function for transforming spatialized output from lat-lon to h3. Also update .gitignore file.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
+.vscode
```
Comment (waltaskew) [inline .gitignore]: You should stick items for your editor setup in a [global gitignore](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c) rather than pull them into every repo you work in.

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,6 @@
...
 apigee/target
 apigee/src/tests
 apigee/src/main/apigee/environments
-cloud_functions/climateiq_data_export_cf/__pycache__
-cloud_functions/climateiq_data_export_cf/.pytest_cache
+**/__pycache__
+**/.pytest_cache
```
Comment (waltaskew) [inline .gitignore]: oughta configure your editor to stick newlines at the end of files, as the laws of unix dictate and to make the github marker go away

File: cloud_functions/convert_h3_cf/main.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import functions_framework
+import pandas as pd
+
+from cloudevents.http import CloudEvent
```
Comment (waltaskew) [inline cloud_functions/convert_h3_cf/main.py]: our [style guide sez](https://google.github.io/styleguide/pyguide.html#22-imports) to only import modules rather than individual classes or functions

[it also sez](https://google.github.io/styleguide/pyguide.html#313-imports-formatting) to group imports by stdlib, third party & sub-package and to keep imports alphabetized within each group.

File: cloud_functions/convert_h3_cf/main.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from google.cloud import storage
+from h3 import h3
+
+H3_LEVEL = 10
+
+
+@functions_framework.cloud_event
+def convert_to_h3(cloud_event: CloudEvent) -> None:
+    """This function is triggered when a new object is created or an existing
+    object is overwritten by the climateiq_data_export_cf Cloud Function.
```
Comment (waltaskew) [inline cloud_functions/convert_h3_cf/main.py]: Is there some documentation around how these two cloud functions work together? And why they're separated into two (why does a first cloud function write outputs for the second to process rather than do said processing in the first function?)
It'd be helpful to write a bit about that in the docstring or link some documentation in notion.

File: cloud_functions/convert_h3_cf/main.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    df = _read_json_to_df(bucket_name, object_name)
+    df["h3"] = df.apply(_lat_lng_to_h3, axis=1)
+    # TODO: Produce different aggregation(s) depending on prediction type.
+    predictions = df.groupby(["h3"]).prediction.agg("mean")
+
+    # TODO: Write output to storage.
+    raise NotImplementedError(predictions.to_json())
+
+
+def _read_json_to_df(bucket_name: str, object_name: str) -> pd.DataFrame:
```
Comment (waltaskew) [inline cloud_functions/convert_h3_cf/main.py]: For passing a data frame from one function to another via file storage, there are some more efficient options than JSON. Are there other readers of these JSON blobs? If it's only Python processes, something like a [pickle](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html)  will be more efficient.

For the feature pipeline, we're using [numpy binary format](https://numpy.org/doc/stable/reference/generated/numpy.save.html)

File: cloud_functions/convert_h3_cf/main.py (hunk 4 above)
Comment (sylmak) [reply cloud_functions/convert_h3_cf/main.py]: Thought about this and agreed it makes sense to combine them.

File: cloud_functions/climateiq_data_export_cf/main.py [hunk 6]
```diff
@@ -205,3 +209,23 @@ def _build_spatialized_model_predictions(
...
+
+    Args:
+        df: A pandas DataFrame containing the results of
+        _build_spatialized_model_predictions.
+    Returns:
+        A Series of h3 indices to the aggregated prediction for the lat/lon center
+        points which fall within each h3 cell.
+    """
+    df["h3"] = df.apply(_lat_lng_to_h3, axis=1)
+    # TODO: Produce different aggregation(s) depending on prediction type.
```
Comment (waltaskew) [inline cloud_functions/climateiq_data_export_cf/main.py]: I think we'll actually always need to do `max` here, because we're interested in max flood height & max temperatures per area.

File: cloud_functions/climateiq_data_export_cf/main.py [hunk 7]
```diff
@@ -205,3 +209,23 @@ def _build_spatialized_model_predictions(
...
+    predictions = df.groupby(["h3"]).prediction.agg("mean")
+    return predictions
+
+
+def _lat_lng_to_h3(row, h3_level=H3_LEVEL):
+    return h3.geo_to_h3(row["lat"], row["lon"], h3_level)
```
Comment (waltaskew) [inline cloud_functions/climateiq_data_export_cf/main.py]: Where are these lat lon columns expected to come from? They won't be present in the model outputs / predictions.

File: cloud_functions/climateiq_data_export_cf/requirements.txt [hunk 8]
```diff
@@ -1,64 +1,14 @@
-affine==2.4.0
```
Comment (waltaskew) [inline cloud_functions/climateiq_data_export_cf/requirements.txt:1]: Did you mean to remove all these requirements?
The requirements file should contain the transitive closure over all requirements.

File: cloud_functions/climateiq_data_export_cf/main.py (hunk 7 above)
Comment (sylmak) [reply cloud_functions/climateiq_data_export_cf/main.py]: The lat lon columns are from the DataFrame generated after running the model outputs and predictions through _build_spatialized_model_predictions.

File: cloud_functions/climateiq_data_export_cf/requirements.txt (hunk 8 above)
Comment (sylmak) [reply cloud_functions/climateiq_data_export_cf/requirements.txt:1]: Got it, reverted.

File: cloud_functions/climateiq_data_export_cf/main.py [hunk 9]
```diff
@@ -205,3 +209,23 @@ def _build_spatialized_model_predictions(
...
+        points which fall within each h3 cell.
+    """
+    df["h3"] = df.apply(_lat_lng_to_h3, axis=1)
+    # TODO: Finalize which aggregation to calculate for each prediction type.
+    predictions = df.groupby(["h3"]).prediction.agg("max")
+    return predictions
+
+
+def _lat_lng_to_h3(row, h3_level=H3_LEVEL):
+    return h3.geo_to_h3(row["lat"], row["lon"], h3_level)
```
Comment (waltaskew) [inline cloud_functions/climateiq_data_export_cf/main.py]: Can also use a lambda for these sorta things
```python
df.apply(lambda row: return h3.geo_to_h3(row["lat"], row["lon"], H3_LEVEL), axis=1)
```

File: cloud_functions/climateiq_data_export_cf/main_test.py [hunk 10]
```diff
@@ -19,16 +19,16 @@ def test_export_model_predictions_invalid_object_name() -> None:
...
 
     with pytest.raises(ValueError) as exc_info:
         main.export_model_predictions(event)
 
     assert "Invalid object name format. Expected 5 components." in str(exc_info.value)
 
 
-@mock.patch.object(main.storage, "Client", autospec=True)
-@mock.patch.object(main.firestore, "Client", autospec=True)
+@mock.patch.object(storage, "Client", autospec=True)
```
Comment (waltaskew) [inline cloud_functions/climateiq_data_export_cf/main_test.py:30]: Why the change here? These should be the equivalent because `storage` is a singleton. I generally prefer the former `@mock.patch.object(main.firestore, "Client", autospec=True)` because the important thing is that you're patching the objects used by the functions inside the `main` module that you're testing.

**Review by waltaskew — APPROVED** [review]
LGTM -- just some small optional things.

File: cloud_functions/climateiq_data_export_cf/main_test.py (hunk 10 above)
Comment (sylmak) [reply cloud_functions/climateiq_data_export_cf/main_test.py:30]: It's a style thing but I've always named the mocked object directly instead of indirectly through the tested module's dependencies. I don't think I've seen this style within google3 before but it's been a few years since I wrote Python extensively. I personally find this more readable because it's easier to understand where the dependencies come from. But if the rest of the ClimateIQ code names the dependencies indirectly, then I can change this.
